testing/src/main/java/com/linecorp/decaton/testing/processor/ProcessorTestSuite.java

```diff
@@ -0,0 +1,318 @@
+        return new Builder<>(rule, parser);
+    }
+
+    /**
+     * Checks processor's basic functionality.
+     * scenario:
+     *   1. Start multiple subscription instances
+     *   2. Produce tasks
+     *   3. When half of tasks are processed, restart subscriptions one by one
+     *   4. Await all tasks be processed
+     *   5. Check process ordering
+     */
+    public void run() {
+        String topicName = rule.admin().createRandomTopic(NUM_PARTITIONS, 3);
+
+        CountDownLatch processLatch = new CountDownLatch(numTasks);
+        CountDownLatch rollingRestartLatch = new CountDownLatch(numTasks / 2);
+
+        List<DecatonProducerRecord<T>> processedTasks = Collections.synchronizedList(new ArrayList<>());
+        Set<T> processedTaskSet = Collections.synchronizedSet(new HashSet<>());
+
+        DecatonProcessor<T> globalProcessor = (context, task) -> {
+            // If a task is retried, ordering is no longer preserved
+            if (context.metadata().retryCount() == 0) {
+                rollingRestartLatch.countDown();
+                processedTasks.add(new DecatonProducerRecord<>(context.key(), task));
+                if (processedTaskSet.add(task)) {
+                    processLatch.countDown();
+                }
+            }
+            context.deferCompletion().completeWith(context.push(task));
+        };
+
+        Supplier<ProcessorSubscription> subscriptionSupplier = () -> {
+            ProcessorsBuilder<T> processorsBuilder =
+                    ProcessorsBuilder.consuming(topicName,
+                                                new ProtocolBuffersDeserializer<>(parser))
+                                     .thenProcess(globalProcessor);
+
+            return TestUtils.subscription(rule.bootstrapServers(),
+                                          configureProcessorsBuilder.apply(processorsBuilder),
+                                          retryConfig,
+                                          propertySuppliers);
+        };
+
+        DecatonClient<T> client = null;
+        ProcessorSubscription[] subscriptions = new ProcessorSubscription[numSubscriptionInstance];
+        try {
+            client = TestUtils.client(topicName, rule.bootstrapServers());
+            for (int i = 0; i < numSubscriptionInstance; i++) {
+                subscriptions[i] = subscriptionSupplier.get();
+            }
+
+            Map<String, List<T>> producedTasks = new HashMap<>();
+            while(tasks.hasNext()) {
+                DecatonProducerRecord<T> record = tasks.next();
+                client.put(record.key(), record.task());
+                List<T> tasks = producedTasks.computeIfAbsent(record.key(), key -> new ArrayList<>());
+                tasks.add(record.task());
+            }
+
+            rollingRestartLatch.await();
+            for (int i = 0; i < numSubscriptionInstance; i++) {
+                subscriptions[i].close();
```

Comment:
For easier investigation when test fails, I think we wanna have more event loggingsaround here

Reply:
ProcessorSubscription's logs not enough?

Reply:
Perhaps, but I think it's much easier to see at which point the problematic thing had happen with associated information on control sequence (restart, stop) that is a contextual information available only at here.